File: typescript/src/wire/crc32c.ts
/**
 * CRC32C (Castagnoli, polynomial 0x1EDC6F41) over a byte slice.
 *
 * Brain checksums every frame header and payload with CRC32C, the same variant
 * hardware accelerates via SSE4.2 `crc32` — picked for its strong error
 * detection and ubiquitous acceleration. The SDK computes it in software so the
 * codec carries no native dependency for a core checksum; the loop is the
 * standard reflected table-driven form.
 *
 * The reflected algorithm processes each byte low-bit-first against a 256-entry
 * table seeded from the reflected polynomial 0x82F63B78, starting and ending
 * with a bitwise inversion (the conventional 0xFFFFFFFF init / final-xor). The
 * check vector `crc32c("123456789") == 0xe3069283` pins this choice.
 */

const POLYNOMIAL_REFLECTED = 0x82f63b78;

// Precompute the byte-wise remainder table once at module load.
const TABLE: Uint32Array = (() => {
  const table = new Uint32Array(256);
  for (let n = 0; n < 256; n++) {
    let c = n;
    for (let k = 0; k < 8; k++) {
      c = c & 1 ? POLYNOMIAL_REFLECTED ^ (c >>> 1) : c >>> 1;
    }
    table[n] = c >>> 0;
  }
  return table;
})();

/** Compute the CRC32C of `bytes`, returned as an unsigned 32-bit integer. */
export function crc32c(bytes: Uint8Array): number {
  let crc = 0xffffffff;
  for (let i = 0; i < bytes.length; i++) {
    crc = TABLE[(crc ^ bytes[i]!) & 0xff]! ^ (crc >>> 8);
  }
  return (crc ^ 0xffffffff) >>> 0;
}

File: python/src/brain_db_sdk/wire/crc32c.py
"""CRC32C (Castagnoli) — the checksum the BRN0 frame header uses.

Pure-Python so the codec has no native/network dependency for a core
function. The algorithm is the reflected Castagnoli CRC (polynomial
0x1EDC6F41, reflected to 0x82F63B78), the same variant hardware SSE4.2
``crc32`` and the server's ``crc32c`` crate compute. A 256-entry table is
built once at import.
"""

from __future__ import annotations

_POLY = 0x82F63B78  # reflected 0x1EDC6F41


def _build_table() -> list[int]:
    table = []
    for n in range(256):
        crc = n
        for _ in range(8):
            crc = (crc >> 1) ^ _POLY if crc & 1 else crc >> 1
        table.append(crc)
    return table


_TABLE = _build_table()


def crc32c(data: bytes) -> int:
    """CRC32C of ``data`` as an unsigned 32-bit integer."""
    crc = 0xFFFFFFFF
    for byte in data:
        crc = (crc >> 8) ^ _TABLE[(crc ^ byte) & 0xFF]
    return crc ^ 0xFFFFFFFF


# Standard check vector pins the polynomial choice: CRC32C("123456789").
# Deliberately a raise, not an assert: `python -O` strips asserts, and this is
# the one line standing between a wrong table and every frame failing CRC on
# the server with no local symptom.
if crc32c(b"123456789") != 0xE3069283:  # pragma: no cover - import-time guard
    msg = "CRC32C check vector failed: the polynomial table is wrong"
    raise RuntimeError(msg)
